pkg/svc/svc.go

```diff
@@ -0,0 +1,486 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package svc
+
+import (
+	"fmt"
+
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/cilium/cilium/pkg/counter"
+	"github.com/cilium/cilium/pkg/k8s"
+	lb "github.com/cilium/cilium/pkg/loadbalancer"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/maps/lbmap"
+	"github.com/cilium/cilium/pkg/metrics"
+	"github.com/cilium/cilium/pkg/service"
+
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "svc")
+
+	updateMetric = metrics.ServicesCount.WithLabelValues("update")
+	deleteMetric = metrics.ServicesCount.WithLabelValues("delete")
+	addMetric    = metrics.ServicesCount.WithLabelValues("add")
+)
+
+// Type is a type of a service.
+type Type string
+
+const (
+	TypeClusterIP = Type("ClusterIP")
+	TypeNodePort  = Type("NodePort")
+)
+
+// Service is a service handler. Its main responsibility is to reflect
+// service-related changes into BPF maps used by datapath BPF programs.
+// The changes can be triggered either by k8s_watcher or directly by
+// API calls to the /services endpoint.
+type Service struct {
+	lock.RWMutex
+
+	svcByHash map[string]*lb.LBSVC
+	svcByID   map[lb.ID]*lb.LBSVC
+
+	backendRefCount counter.StringCounter
+	backendByHash   map[string]*lb.LBBackEnd
+}
+
+// NewService creates a new instance of the service handler.
+func NewService() *Service {
+	return &Service{
+		svcByHash:       map[string]*lb.LBSVC{},
+		svcByID:         map[lb.ID]*lb.LBSVC{},
+		backendRefCount: counter.StringCounter{},
+		backendByHash:   map[string]*lb.LBBackEnd{},
+	}
+}
+
+// InitMaps opens or creates BPF maps used by services.
+//
+// If restore is set to false, entries of the maps are removed.
+func (s *Service) InitMaps(ipv6, ipv4, restore bool) error {
+	s.Lock()
+	defer s.Unlock()
+
+	// The following two calls can be removed in v1.8+.
+	if err := bpf.UnpinMapIfExists("cilium_lb6_rr_seq_v2"); err != nil {
+		return nil
+	}
+	if err := bpf.UnpinMapIfExists("cilium_lb4_rr_seq_v2"); err != nil {
+		return nil
+	}
+
+	toOpen := []*bpf.Map{}
+	toDelete := []*bpf.Map{}
+	if ipv6 {
+		toOpen = append(toOpen, lbmap.Service6MapV2, lbmap.Backend6Map, lbmap.RevNat6Map)
+		if !restore {
+			toDelete = append(toDelete, lbmap.Service6MapV2, lbmap.Backend6Map, lbmap.RevNat6Map)
+		}
+	}
+	if ipv4 {
+		toOpen = append(toOpen, lbmap.Service4MapV2, lbmap.Backend4Map, lbmap.RevNat4Map)
+		if !restore {
+			toDelete = append(toDelete, lbmap.Service4MapV2, lbmap.Backend4Map, lbmap.RevNat4Map)
+		}
+	}
+
+	for _, m := range toOpen {
```

Comment:
Shouldn't we delete first? If `restore` is false won't we "OpenOrCreate" a map and then delete it?

Reply:
No, we cannot delete entries if the maps haven't been opened.

Previously, we did "OpenOrCreate" (https://github.com/cilium/cilium/blob/master/daemon/datapath.go#L575), and only afterwards "DeleteAll" (https://github.com/cilium/cilium/blob/master/daemon/datapath.go#L601).

Reply:
ah it's `DeleteAll` I read it as `DeleteMap`